Refactor/core reaction

akaviaLab opened the pull request:
* [X] description of feature/fix
Updated reaction.py and test_core_reaction.py to be Python 3.6+ compliant.
I didn't fully fix memo in deepcopy, because I don't know what it is supposed to be and/or do.

Commit:
update reaction.py to Python 3.6+

Commit:
update test_core_reaction.py to Python 3.6+
minor fix to reaction.py

Commit:
black contradicts with flake8

Commit:
fix black contradicts with flake8 configuration

synchon commented on src/cobra/test/test_core/test_core_reaction.py:
```diff
@@ -44,7 +44,7 @@ def test_gpr() -> None:
 
 
 def test_gpr_uppercase():
```
Please add the return type annotation as `None`.

synchon commented on src/cobra/test/test_core/test_core_reaction.py:
```diff
@@ -54,7 +54,10 @@ def test_gpr_uppercase():
 
 @pytest.mark.parametrize("input_gpr", ["(a1 or a2", "(forT or "])
 def test_gpr_malformed(input_gpr):
```
Please add the argument type annotation as `str` and return type annotation as `None`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1,20 +1,33 @@
+    from optlang.interface import Variable
+    from cobra import Model, Solution
+    from cobra.summary import ReactionSummary
+    from pandas import DataFrame
```
Please use `import pandas as pd` and then refer `DataFrame` as `pd.DataFrame` so as to maintain the standard used by `pandas` community and also make it easy for us to understand foreign imports.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1,20 +1,33 @@
+    from cobra.summary import ReactionSummary
+    from pandas import DataFrame
 
 from cobra.core.configuration import Configuration
```
Please use relative imports (wherever applicable) to refer to internal imports, for example: `from .configuration import Configuration` .

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -62,20 +71,30 @@ class Reaction(Object):
     ----------
     id : string
         The identifier to associate with this reaction
```
Please mention the default value for the parameter in the docstring as:
`The identifier to associate with this reaction (default None).`

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -62,20 +71,30 @@ class Reaction(Object):
     id : string
         The identifier to associate with this reaction
-    name : string
+    name : string, optional
```
Please adapt the previous comment for this docstring as well. Also, prefer `str` for docstring as well and keep it consistent throughout.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -62,20 +71,30 @@ class Reaction(Object):
+    name : string, optional
         A human readable name for the reaction
-    subsystem : string
+    subsystem : string, optional
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -62,20 +71,30 @@ class Reaction(Object):
+    subsystem : string, optional
         Subsystem where the reaction is meant to occur
-    lower_bound : float
+    lower_bound : float, optional
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -62,20 +71,30 @@ class Reaction(Object):
+    lower_bound : float, optional
         The lower flux bound
-    upper_bound : float
+    upper_bound : float, optional
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -62,20 +71,30 @@ class Reaction(Object):
-        Object.__init__(self, id, name)
+        self,
+        id: Optional[str] = None,
+        name: Optional[str] = "",
```
Please only use `str` for the type annotation.

akaviaLab replied:
You mean not Optional for empty strings?
Another maintainer said that default of None should be Optional. Do you agree?

synchon replied:
If the default value is `""` (empty string), the type should be `str`. And, if the default value is `None`, then the type becomes `Optional[...]`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -62,20 +71,30 @@ class Reaction(Object):
+        self,
+        id: Optional[str] = None,
+        name: Optional[str] = "",
+        subsystem: Optional[str] = "",
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -62,20 +71,30 @@ class Reaction(Object):
+        id: Optional[str] = None,
+        name: Optional[str] = "",
+        subsystem: Optional[str] = "",
+        lower_bound: Optional[float] = 0.0,
```
Please adapt the previous comment for this type annotation as well.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -84,7 +103,7 @@ def __init__(
         # A dictionary of metabolites and their stoichiometric coefficients in
         # this reaction.
-        self._metabolites = {}
+        self._metabolites = dict()
```
Please use literal syntax like so: `{}` in place of `dict()` .

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -98,10 +117,19 @@ def __init__(
+        ----------
+        value: str
+            string that represents the id.
+        """
```
As the function can raise exception, please add that in the docstring as well.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -111,31 +139,32 @@ def _set_id_with_model(self, value):
 
     @property
-    def reverse_id(self):
+    def reverse_id(self) -> str:
```
Please add a detailed docstring for this method.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -111,31 +139,32 @@ def _set_id_with_model(self, value):
     @property
-    def flux_expression(self):
-        """Forward flux expression
+    def flux_expression(self) -> Union["Variable", None]:
```
Please use `Optional["Variable"]` for the return type.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -111,31 +139,32 @@ def _set_id_with_model(self, value):
     @property
-    def forward_variable(self):
-        """An optlang variable representing the forward flux
+    def forward_variable(self) -> Union["Variable", None]:
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -149,25 +178,23 @@ def forward_variable(self):
     @property
-    def reverse_variable(self):
-        """An optlang variable representing the reverse flux
+    def reverse_variable(self) -> Union["Variable", None]:
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -149,25 +178,23 @@ def forward_variable(self):
-    def objective_coefficient(self):
-        """Get the coefficient for this reaction in a linear
-        objective (float)
+    def objective_coefficient(self) -> float:
```
Please document the return type in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -177,29 +204,37 @@ def objective_coefficient(self):
 
     @objective_coefficient.setter
-    def objective_coefficient(self, value):
+    def objective_coefficient(self, value: float) -> None:
```
Please add a docstring including Parameters and Raises sections.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -177,29 +204,37 @@ def objective_coefficient(self):
             set_objective(self.model, {self: value}, additive=True)
 
-    def __copy__(self):
+    def __copy__(self) -> "Reaction":
```
Please add the Returns section in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -177,29 +204,37 @@ def objective_coefficient(self):
 
-    def __deepcopy__(self, memo):
+    # Unclear what memo should be
+    def __deepcopy__(self, memo) -> "Reaction":
```
Same as above.

akaviaLab replied:
Done. Can you help me figure out what memo should be?

Midnighter replied:
It's a `dict` https://docs.python.org/3/library/copy.html#copy.deepcopy

```suggestion
    def __deepcopy__(self, memo: dict) -> "Reaction":
```

synchon replied:
According to https://docs.python.org/3/library/copy.html , `memo` should be a dictionary, so the type can be `Dict`. A generic `Dict` without key and value type is fine in my opinion.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -177,29 +204,37 @@ def objective_coefficient(self):
 
     @staticmethod
-    def _check_bounds(lb, ub):
+    def _check_bounds(lb: Union[int, float], ub: Union[int, float]) -> None:
```
As the input types for `lower_bound` and `upper_bound` are taken as `float`, we can use `float` for the argument types as well. Also, please add a Raises section in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -177,29 +204,37 @@ def objective_coefficient(self):
+    def update_variable_bounds(self) -> None:
+        """Update and correct variable bounds.
+
+        Sets the forward_variable and reverse_varaible bounds based on lower and
```
A small typo: `variable` in place of `varaible`

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -224,8 +259,8 @@ def update_variable_bounds(self):
     @property
-    def lower_bound(self):
-        """Get or set the lower bound
+    def lower_bound(self) -> Union[int, float]:
```
As earlier, please make the return type as `float` and also add a Returns section to the docstring. This method only gets the value, so I believe you can move the sets directive in the docstring to the setter method.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -239,15 +274,15 @@ def lower_bound(self):
     @lower_bound.setter
     @resettable
-    def lower_bound(self, value):
+    def lower_bound(self, value: Union[int, float]) -> None:
```
Please make the argument type as `float` and also add a docstring with the necessary information.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -239,15 +274,15 @@ def lower_bound(self):
     @property
-    def upper_bound(self):
-        """Get or set the upper bound
+    def upper_bound(self) -> Union[int, float]:
```
Same comment as for the lower bound case.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -261,15 +296,15 @@ def upper_bound(self):
     @upper_bound.setter
     @resettable
-    def upper_bound(self, value):
+    def upper_bound(self, value: Union[int, float]) -> None:
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -261,15 +296,15 @@ def upper_bound(self):
     @property
-    def bounds(self):
-        """Get or set the bounds directly from a tuple
+    def bounds(self) -> Tuple[Union[float, int], Union[float, int]]:
```
Please adapt the return type and also document it in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -282,7 +317,7 @@ def bounds(self):
     @bounds.setter
     @resettable
-    def bounds(self, value):
+    def bounds(self, value: Union[Tuple, list]) -> None:
```
Please add a docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -291,9 +326,9 @@ def bounds(self, value):
 
     @property
-    def flux(self):
+    def flux(self) -> Union[int, float]:
```
I believe flux value is always returned as `float`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -333,24 +368,20 @@ def flux(self):
 
     @property
-    def reduced_cost(self):
+    def reduced_cost(self) -> Union[int, float]:
```
I think reduced cost is always returned as `float`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -389,23 +420,20 @@ def reduced_cost(self):
     @property
-    def metabolites(self):
+    def metabolites(self) -> Dict:
+        """Get a dictionary of metabolites and coefficients."""
```
Please add a return type in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -504,12 +529,12 @@ def gene_name_reaction_rule(self):
 
     @property
-    def gpr(self):
+    def gpr(self) -> GPR:
```
Please add a Returns section to the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -538,8 +563,8 @@ def functional(self):
     @property
-    def x(self):
-        """The flux through the reaction in the most recent solution.
+    def x(self) -> Union[int, float]:
```
The return type can be `float`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -548,8 +573,8 @@ def x(self):
     @property
-    def y(self):
-        """The reduced cost of the reaction in the most recent solution.
+    def y(self) -> Union[int, float]:
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -558,33 +583,41 @@ def y(self):
 
     @reversibility.setter
-    def reversibility(self, value):
+    def reversibility(self, value: bool) -> None:
```
Please add a docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -558,33 +583,41 @@ def y(self):
     @property
-    def model(self):
-        """returns the model the reaction is a part of"""
+    def model(self) -> "Model":
```
Please add a Returns section in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -593,8 +626,8 @@ def _update_awareness(self):
 
-    def remove_from_model(self, remove_orphans=False):
-        """Removes the reaction from a model.
+    def remove_from_model(self, remove_orphans: Optional[bool] = False) -> None:
```
Please make the type for `remove_orphans` as `bool`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -609,15 +642,16 @@ def remove_from_model(self, remove_orphans=False):
 
-    def delete(self, remove_orphans=False):
-        """Removes the reaction from a model.
+    def delete(self, remove_orphans: Optional[bool] = False) -> None:
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -631,8 +665,10 @@ def delete(self, remove_orphans=False):
 
-    def __setstate__(self, state):
-        """Probably not necessary to set _model as the cobra.Model that
+    def __setstate__(self, state: Dict) -> None:
```
Please add a Parameters section to the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -653,14 +689,14 @@ def __setstate__(self, state):
 
-    def copy(self):
-        """Copy a reaction
+    def copy(self) -> "Reaction":
```
Please add a Returns section to the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -682,13 +718,19 @@ def copy(self):
+        the first reaction.
+
+        Does not modify in place.
+
```
Please add a Parameters section here.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -701,38 +743,82 @@ def __add__(self, other):
+        and. All other attributes (i.e. reaction bounds) will match those of
+        the first reaction.
+
+        Modifies in place.
```
Please add Parameters and Returns sections here.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -701,38 +743,82 @@ def __add__(self, other):
+        the first reaction.
+
+        Does not modify in place. The name will still be that of the first reaction.
+
```
Please add a Parameters section here.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -701,38 +743,82 @@ def __add__(self, other):
+        the first reaction.
+
+        Modifies in place and changes the original reaction.
+
```
Please add a Parameters section here.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -701,38 +743,82 @@ def __add__(self, other):
 
-    def __imul__(self, coefficient):
-        """Scale coefficients in a reaction by a given value
+    def __imul__(self, coefficient: Union[int, float]) -> "Reaction":
```
Please enforce `float` for the type annotation and the docstring as well.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -751,24 +837,43 @@ def __imul__(self, coefficient):
         return self
 
-    def __mul__(self, coefficient):
+    def __mul__(self, coefficient: Union[int, float]) -> "Reaction":
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -751,24 +837,43 @@ def __imul__(self, coefficient):
 
     @property
-    def reactants(self):
+    def reactants(self) -> List[Metabolite]:
```
Please add a Returns section to the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -751,24 +837,43 @@ def __imul__(self, coefficient):
     @property
-    def products(self):
-        """Return a list of products for the reaction"""
+    def products(self) -> List[Metabolite]:
```
Please add a Returns section to the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -751,24 +837,43 @@ def __imul__(self, coefficient):
-        Return the stoichiometric coefficient of a metabolite.
+    def get_coefficient(
+        self, metabolite_id: Union[str, Metabolite]
+    ) -> Union[int, float]:
```
I believe the return is always a `float`, can you please check it? Also, please add a Returns section to the docstring.

akaviaLab replied:
I'm not sure, but I made it a float.

synchon replied:
Okay let's keep it like that for now, and if required, we can change it later.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -781,20 +886,32 @@ def get_coefficient(self, metabolite_id):
-        Return the stoichiometric coefficients for a list of metabolites.
+    def get_coefficients(
+        self, metabolite_ids: Iterable[Union[str, Metabolite]]
+    ) -> Iterator[Union[int, float]]:
```
I think the return type can be `Iterable[float]`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -781,20 +886,32 @@ def get_coefficient(self, metabolite_id):
-    def add_metabolites(self, metabolites_to_add, combine=True, reversibly=True):
+    def add_metabolites(
+        self,
+        metabolites_to_add: Dict[Metabolite, Union[int, float]],
```
Please enforce type `float` for the values.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -781,20 +886,32 @@ def get_coefficient(self, metabolite_id):
+    def add_metabolites(
+        self,
+        metabolites_to_add: Dict[Metabolite, Union[int, float]],
+        combine: Optional[bool] = True,
```
Please make the type as `bool` and document the optional behaviour in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -781,20 +886,32 @@ def get_coefficient(self, metabolite_id):
+        self,
+        metabolites_to_add: Dict[Metabolite, Union[int, float]],
+        combine: Optional[bool] = True,
+        reversibly: Optional[bool] = True,
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -914,7 +1030,12 @@ def add_metabolites(self, metabolites_to_add, combine=True, reversibly=True):
-    def subtract_metabolites(self, metabolites, combine=True, reversibly=True):
+    def subtract_metabolites(
+        self,
+        metabolites: Dict[Metabolite, Union[int, float]],
```
The type for the value can be `float`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -914,7 +1030,12 @@ def add_metabolites(self, metabolites_to_add, combine=True, reversibly=True):
+    def subtract_metabolites(
+        self,
+        metabolites: Dict[Metabolite, Union[int, float]],
+        combine: Optional[bool] = True,
```
Please make the type as `bool` and document the optional behaviour in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -914,7 +1030,12 @@ def add_metabolites(self, metabolites_to_add, combine=True, reversibly=True):
+        self,
+        metabolites: Dict[Metabolite, Union[int, float]],
+        combine: Optional[bool] = True,
+        reversibly: Optional[bool] = True,
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -950,18 +1071,20 @@ def subtract_metabolites(self, metabolites, combine=True, reversibly=True):
-    def reaction(self):
-        """Human readable reaction string"""
+    def reaction(self) -> str:
+        """Return Human readable reaction string."""
```
Please add a Returns section in the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -950,18 +1071,20 @@ def subtract_metabolites(self, metabolites, combine=True, reversibly=True):
     @reaction.setter
-    def reaction(self, value):
-        return self.build_reaction_from_string(value)
+    def reaction(self, value: str):
```
Please add a docstring and the return type annotation.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -950,18 +1071,20 @@ def subtract_metabolites(self, metabolites, combine=True, reversibly=True):
-    def build_reaction_string(self, use_metabolite_names=False):
-        """Generate a human readable reaction string"""
+    def build_reaction_string(
+        self, use_metabolite_names: Optional[bool] = False
```
Please use `bool` as the type for `use_metabolite_names`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -950,18 +1071,20 @@ def subtract_metabolites(self, metabolites, combine=True, reversibly=True):
+    def build_reaction_string(
+        self, use_metabolite_names: Optional[bool] = False
+    ) -> str:
+        """Generate a human readable reaction string."""
```
Please add Parameters and Returns section to the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -988,37 +1111,40 @@ def format(number):
+        dict
+            a dict of {element: amount} for unbalanced elements.
+            "charge" is treated as an element in this dict
+            This should be empty for balanced reactions.
```
Please add a Raises section here.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -988,37 +1111,40 @@ def format(number):
 
-    def check_mass_balance(self):
-        """Compute mass and charge balance for the reaction
+    def check_mass_balance(self) -> Dict[str, Union[int, float]]:
```
I think the value for the return can be of type `float`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -988,37 +1111,40 @@ def format(number):
-    def compartments(self):
-        """lists compartments the metabolites are in"""
+    def compartments(self) -> Set:
+        """Return set of compartments the metabolites are in."""
```
Please add a Returns section to the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -988,37 +1111,40 @@ def format(number):
 
-    def get_compartments(self):
-        """lists compartments the metabolites are in"""
+    def get_compartments(self) -> list:
```
Maybe add a Returns section to the docstring? Can be kept as it is as it has a deprecation warning on it.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1041,20 +1167,20 @@ def _dissociate_gene(self, cobra_gene):
-    ):
-        """Builds reaction from reaction equation reaction_str using parser
+        reaction_str: str,
+        verbose: Optional[bool] = True,
```
Please use `bool` for the type.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1041,20 +1167,20 @@ def _dissociate_gene(self, cobra_gene):
+        fwd_arrow: Optional[AnyStr] = None,
+        rev_arrow: Optional[AnyStr] = None,
+        reversible_arrow: Optional[AnyStr] = None,
+        term_split: Optional[str] = "+",
```
Please use `str` for the type.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1070,11 +1196,11 @@ def build_reaction_from_string(
+        reversible_arrow : AnyStr
             for reversible reaction arrows
         term_split : string
             dividing individual metabolite entries
```
Please add necessary information for a Raises section here.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1041,20 +1167,20 @@ def _dissociate_gene(self, cobra_gene):
-        """Builds reaction from reaction equation reaction_str using parser
+        reaction_str: str,
+        verbose: Optional[bool] = True,
+        fwd_arrow: Optional[AnyStr] = None,
```
I believe `re.compile()` generates a regular expression object and `AnyStr` is not accurate for that.

akaviaLab replied:
The code uses re.compile(re.escape(fwd_arrow)) (see line 1213 below, line 1501 in the revised code, line 1087 in the original code). re.escape accepts a pattern, either str or bytes (binary str). AnyStr is equal to Union[str, bytes].
I can also set it as str, but it definitely shouldn't be Pattern. Try re.escape(re.compile('-->')) and you'll get an error.

synchon replied:
Okay, yes you are correct. My bad, I missed the usage part. Let's keep it `AnyStr`.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1041,20 +1167,20 @@ def _dissociate_gene(self, cobra_gene):
+        reaction_str: str,
+        verbose: Optional[bool] = True,
+        fwd_arrow: Optional[AnyStr] = None,
+        rev_arrow: Optional[AnyStr] = None,
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1041,20 +1167,20 @@ def _dissociate_gene(self, cobra_gene):
+        verbose: Optional[bool] = True,
+        fwd_arrow: Optional[AnyStr] = None,
+        rev_arrow: Optional[AnyStr] = None,
+        reversible_arrow: Optional[AnyStr] = None,
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1186,42 +1316,38 @@ def summary(self, solution=None, fva=None):
-        return "{id}: {stoichiometry}".format(
-            id=self.id, stoichiometry=self.build_reaction_string()
-        )
+    def __str__(self) -> str:
```
Please add a Returns section for the docstring.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1186,42 +1316,38 @@ def summary(self, solution=None, fva=None):
 
-    def _repr_html_(self):
-        return """
+    def _repr_html_(self) -> str:
```
Same as above.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1186,42 +1316,38 @@ def summary(self, solution=None, fva=None):
             </tr><tr>
                 <td><strong>Memory address</strong></td>
-                <td>{address}</td>
+                <td>{f"0x0{id(self):x}"}</td>
```
Please make it as: `{id(self):#x}`.

synchon reviewed (changes requested):
I have tried my best to go through everything. As the file is large, I might have missed a couple of things, so feel free to add those enhancements.

Commit:
applied code review suggestions to reaction.py

Commit:
applied code review suggestions to test_core_reaction.py

akaviaLab reviewed (commented):
Thank you very much for your detailed comments. I tried to respond to all of them.
Can you please help me with memo (in deepcopy)? I can't figure out what it is supposed to be?

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -60,22 +69,32 @@ class Reaction(Object):
+    lower_bound : float, default 0
+        The lower flux bound (default 0).
+    upper_bound : float, optional, default None
+        The upper flux bound (default None).
```
This is a new style and not in line with numpydoc https://numpydoc.readthedocs.io/en/latest/format.html#parameters, default values should be in the description.

```suggestion
    id : str or None, optional
        The identifier to associate with this reaction (default None).
    name : str, optional
        A human readable name for the reaction (default "").
    subsystem : str, optional
        Subsystem where the reaction is meant to occur (default "").
    lower_bound : float, optional
        The lower flux bound (default 0).
    upper_bound : float or None, optional
        The upper flux bound (default None).
```

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -149,57 +190,121 @@ def forward_variable(self):
+
+        Parameters
+        ----------
+        memo
```
```suggestion
        memo: dict
            Automatically passed parameter.
```

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -224,65 +329,117 @@ def update_variable_bounds(self):
     @bounds.setter
     @resettable
-    def bounds(self, value):
+    def bounds(self, value: Union[Tuple[float, float], List[float]]) -> None:
```
It would be more concise to say that a [Sequence](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) interface needs to be supported.

```suggestion
    def bounds(self, value: Sequence[float]) -> None:
```

but if you want to highlight that the usual case is a pair that's fine, too:

```suggestion
    def bounds(self, value: Union[Tuple[float, float], Sequence[float]]) -> None:
```

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -224,65 +329,117 @@ def update_variable_bounds(self):
+
+        Parameters
+        ----------
+        value: tuple or list
```
```suggestion
        value: tuple or sequence
```

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -333,28 +495,29 @@ def flux(self):
         # Due to below all-catch, which sucks, need to reraise these.
         except (RuntimeError, OptimizationError) as err:
-            raise_with_traceback(err)
+            raise err.with_traceback()
```
More modern syntax:

```suggestion
            raise
```

one could also consider raising some kind of cobrapy error.

akaviaLab replied:
I read [Raising exceptions](https://python-future.org/compatible_idioms.html#raising-exceptions) and it seems to indicate that Python 3 requires err.with_traceback()
Should I import traceback or something?

Midnighter replied:
Take a look at the official tutorial. I think the futures docs don't give the most recent impression since that is focused on compatible 2 & 3 syntax.

https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement

> If no expressions are present, [raise](https://docs.python.org/3/reference/simple_stmts.html#raise) re-raises the exception that is currently being handled, which is also known as the active exception. If there isn’t currently an active exception, a [RuntimeError](https://docs.python.org/3/library/exceptions.html#RuntimeError) exception is raised indicating that this is an error.

synchon commented:
Please rebase it on the latest `devel` after you push your final changes.

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -389,23 +552,29 @@ def reduced_cost(self):
         # Due to below all-catch, which sucks, need to reraise these.
         except (RuntimeError, OptimizationError) as err:
-            raise_with_traceback(err)
+            raise err.with_traceback()
```
As above

```suggestion
            raise
```

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -538,53 +715,97 @@ def functional(self):
     @property
-    def x(self):
-        """The flux through the reaction in the most recent solution.
+    def x(self) -> float:
```
These have been deprecated for some time. Maybe it's time to remove them?

synchon replied:
Maybe a separate PR to get rid of deprecations?

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -538,53 +715,97 @@ def functional(self):
     @property
-    def y(self):
-        """The reduced cost of the reaction in the most recent solution.
+    def y(self) -> float:
```
Time to remove `.y`?

Midnighter replied:
Separate PR

Midnighter commented on src/cobra/core/reaction.py:
```diff
@@ -1066,19 +1472,23 @@ def build_reaction_from_string(
+        reaction_str : str
             a string containing a reaction formula (equation)
-        verbose: bool
+        verbose: bool, default True
             setting verbosity of function
```
```suggestion
        verbose: bool, optional
            Set the verbosity of the function (default `True`).
```

Midnighter reviewed (changes requested):
Couple more changes otherwise it looks great. Thanks for the big update 🙂

Commit:
applied code suggestoins

akaviaLab commented:
> Please rebase it on the latest `devel` after you push your final changes.

Resolved conflicts, which failed.
Rebasing and force pushing to avoid having 6 copies of the commits

Commit:
black on the merged versions

akaviaLab committed:
fix raise_with_traceback

https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement indicates that 
raise err.with_traceback() is not necessary, despite what future documentation says.

Co-authored-by: Moritz E. Beber <midnighter@posteo.net>

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1,44 +1,52 @@
-from cobra.util.solver import (
+if TYPE_CHECKING:
+    from optlang.interface import Variable
+    from . import Model, Solution
```
Apologies for not pointing out earlier: for imports under `TYPE_CHECKING`, you will need the full import path like so: `from cobra import Model, Solution`.

akaviaLab replied:
Is this a stylistic choice or a syntax issue?
Because it seems to work like this.

synchon replied:
So, static type checkers like `mypy` need to import `cobra` just like `pytest` does for testing. `TYPE_CHECKING` is respected by static type checkers. From my experience of using `mypy`, it reports error. Since, we don't have a static type checking in CI as of now, it will pass the checks. It will be good to have it for future.

akaviaLab replied:
Done

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1,44 +1,52 @@
+if TYPE_CHECKING:
+    from optlang.interface import Variable
+    from . import Model, Solution
+    from ..summary import ReactionSummary
```
`from cobra.summary import ReactionSummary`

akaviaLab replied:
Same question as above.

synchon replied:
See above comment.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -61,22 +71,32 @@ class Reaction(Object):
-        The lower flux bound
-    upper_bound : float
-        The upper flux bound
+        The lower flux bound (default 0).
```
`... (default 0.0).`

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -150,57 +192,122 @@ def forward_variable(self):
+        Returns
+        ------
+        float
+            Linear coefficient if this reaction has any, or 0 otherwise.
```
... 0.0 otherwise.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -458,31 +629,28 @@ def _update_genes_from_gpr(self, new_gene_names: set = None) -> None:
         return self._gpr.to_string()
 
     @gene_reaction_rule.setter
     def gene_reaction_rule(self, new_rule: str):
```
Please add the return type.

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -812,14 +1162,22 @@ def add_metabolites(self, metabolites_to_add, combine=True, reversibly=True):
         reversibly : bool
             Whether to add the change to the context to make the change
-            reversibly or not (primarily intended for internal use).
+            reversibly or not (primarily intended for internal use). Default is True.
```
... (default True).

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -937,14 +1299,14 @@ def subtract_metabolites(self, metabolites, combine=True, reversibly=True):
             False causes the coefficient to be replaced.
 
-        reversibly : bool
+        reversibly : bool, defalt True
```
Please move the `default True` to the description as `... (default True).`

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -1069,19 +1482,23 @@ def build_reaction_from_string(
+        rev_arrow : AnyStr, optional
+            Str or bytes that encode backward irreversible reaction arrows (default None).
+        reversible_arrow : AnyStr, optional
+            Str or bytes that encode reversible reaction arrows (defualt None).
```
A small typo: `... (default True).`

synchon commented on src/cobra/test/test_core/test_core_reaction.py:
```diff
@@ -53,8 +53,11 @@ def test_gpr_uppercase():
+def test_gpr_malformed(input_gpr: str) -> None:
+    """Test ability to deal with malformed GPR.
+
+    Malformed GPR strings will lead to empty GPRs with no genes.
```
Please add a Parameters section here.

synchon reviewed (changes requested):
A few more changes.

Commit:
more corrections as requested

Commit:
fix for static type checking

synchon commented:
Thank you for your contribution!

Midnighter commented:
I don't follow exactly why you made those changes in the last commit.

akaviaLab commented:
Hi @Midnighter - The fist change is based on the conversation linked in https://github.com/opencobra/cobrapy/pull/1171#discussion_r825337513, where I understood @synchon would like these last changes to happen.
The change to err is because my linter was complaining that it was not used.
The third change is just line formatting.

Midnighter commented:
That makes sense. Thank you.
